From aaccbdedfc56f437e7a4f6ce6c33719e58e125e8 Mon Sep 17 00:00:00 2001
From: Martin Blumenstingl <martin.blumenstingl@googlemail.com>
Date: Sat, 25 Dec 2021 04:22:53 +0100
Subject: [PATCH 09/98] net: stmmac: dwmac-meson: Manage the "ethernet" clock

Meson6 and Meson8 (both use the same glue registers on top of the DWMAC
IP) have a dedicated Ethernet clock. For RMII mode the SoC has an input
for an external RMII reference clock signal (which can be provided by
either the PHY or an external oscillator). This clock needs to run at
50MHz because the additional glue registers can divide by 2 - to achieve
25MHz for 100Mbit/s line speed, or 20 - to achieve 2.5MHz for 10Mbit/s
line speed.

Set the correct frequency for this clock and enable it during init. Also
enable the ETHMAC_DIV_EN bit which enables the divider in the glue
registers, based on the Ethernet clock input.

Signed-off-by: Martin Blumenstingl <martin.blumenstingl@googlemail.com>

[Rebased for kernel 6.12.96: this kernel version's struct
plat_stmmacenet_data still has the old .fix_mac_speed callback instead
of .set_clk_tx_rate (that rename happened in a later, unrelated stmmac
core refactor), so meson6_dwmac_fix_mac_speed/its assignment are kept
as-is instead of being renamed. Also .init()/.exit() are called by
stmmac_platform.c on this kernel version as
`plat->init(pdev, plat->bsp_priv)` with a struct platform_device *,
not struct device * (that too changed later upstream), so
meson6_dwmac_init/meson6_dwmac_exit take a struct platform_device *pdev
parameter here.]
---
 .../net/ethernet/stmicro/stmmac/dwmac-meson.c | 30 +++++++++++++++++++
 1 file changed, 30 insertions(+)

diff --git a/drivers/net/ethernet/stmicro/stmmac/dwmac-meson.c b/drivers/net/ethernet/stmicro/stmmac/dwmac-meson.c
--- a/drivers/net/ethernet/stmicro/stmmac/dwmac-meson.c
+++ b/drivers/net/ethernet/stmicro/stmmac/dwmac-meson.c
@@ -5,6 +5,7 @@
  * Copyright (C) 2014 Beniamino Galvani <b.galvani@gmail.com>
  */

+#include <linux/clk.h>
 #include <linux/device.h>
 #include <linux/ethtool.h>
 #include <linux/io.h>
@@ -15,6 +16,8 @@

 #include "stmmac_platform.h"

+#define PREG_ETHERNET_ADDR0_DIV_EN	BIT(0)
+
 /* divides the input clock by 20 (= 0x0) or 2 (= 0x1) */
 #define PREG_ETHERNET_ADDR0_SPEED_100	BIT(1)

@@ -42,11 +45,28 @@
 	writel(val, dwmac->reg);
 }

+static int meson6_dwmac_init(struct platform_device *pdev, void *priv)
+{
+	struct meson_dwmac *dwmac = priv;
+
+	writel(readl(dwmac->reg) | PREG_ETHERNET_ADDR0_DIV_EN, dwmac->reg);
+
+	return 0;
+}
+
+static void meson6_dwmac_exit(struct platform_device *pdev, void *priv)
+{
+	struct meson_dwmac *dwmac = priv;
+
+	writel(readl(dwmac->reg) & ~PREG_ETHERNET_ADDR0_DIV_EN, dwmac->reg);
+}
+
 static int meson6_dwmac_probe(struct platform_device *pdev)
 {
 	struct plat_stmmacenet_data *plat_dat;
 	struct stmmac_resources stmmac_res;
 	struct meson_dwmac *dwmac;
+	struct clk *clk;
 	int ret;

 	ret = stmmac_get_platform_resources(pdev, &stmmac_res);
@@ -65,7 +85,14 @@
 	if (IS_ERR(dwmac->reg))
 		return PTR_ERR(dwmac->reg);

+	clk = devm_clk_get_optional_enabled_with_rate(&pdev->dev, "ethernet",
+						      50 * 1000 * 1000);
+	if (IS_ERR(clk))
+		return PTR_ERR(clk);
+
 	plat_dat->bsp_priv = dwmac;
+	plat_dat->init = meson6_dwmac_init;
+	plat_dat->exit = meson6_dwmac_exit;
 	plat_dat->fix_mac_speed = meson6_dwmac_fix_mac_speed;

 	return stmmac_dvr_probe(&pdev->dev, plat_dat, &stmmac_res);
